### DISCUSSION TITLE
Access Astro DB schema and data from an Astro integration

### DISCUSSION THREAD

**Discussion by lorenzolewis** [withastro/roadmap#914] [Proposal]
# Summary

Expose the ability (possibly via an integration hook) to access data and schemas through Astro DB from within another integration.

# Background & Motivation

We have the `astro:db:setup` hook that we can use to extend a database with additional configuration and seed data. But now (to the best of my knowledge) it isn't possible to then get access to the data that is then populated (either through seeding, a local file, or through Astro Studio).

This limits the use cases for integrations to provide utilities for the injected tables and respective data if provided.

# Goals

- Allow integration developers to access data that would normally be available via `import * from astro:db`
- Allow access to read data (and possibly write) in an integration hook

# Example

Below is a pseudo-code example of how this could be used to power an i18n utility backed by Astro DB:

```ts
// integration definition

const i18n = new i18nTool();

return {
  hooks: {
    "astro:db:setup": ({ extendDb }) => {
      extendDb({ configEntrypoint: resolve("./db/config") });
    },
    "astro:db:ready": ({ readDb }) => {
      const data = await readDb(TranslationSelectQueryHere);
      i18n.loadTranslations(data)
    }
  }
}
```

A difficulty would be how to get access to data in a type-safe manner if at possible for use in querying the database. This gets into virtual module territory and I'm afraid my knowledge falls short in that area to offer any suggestions.

**Comment by lorenzolewis** [comment]
Another approach could be to locally define the tables like this:

This would require re-exporting utilities through `@astrojs/db` (or similar) so that the Vite virtual modules could be circumnavigated. 

```ts
import { defineTable } from "@astrojs/db"
const Translations = defineTable({foobar})
return {
  hooks: {
    "astro:db:setup": ({ extendDb }) => {
      extendDb({ addTables: [Translations] });
    },
    "astro:db:ready": ({ db }) => {
      const data = await db.select().from(Translations)
      i18n.loadTranslations(data)
    }
  }
}
```

**Reply by lorenzolewis** [reply]
I did a little bit of experimenting in code to mock this up today. Here's a few notes from it:

I believe we would need to define a DB and not a table here for cases where there would be relationships
   I'm not sure how the fits with the whole architecture of Sqlite and Astro Studio
- Would it be allowed to have multiple DBs? Would it work to merge the tables together?
- The current method that's used in with `configEntrypoints` is to merge the tables: https://github.com/withastro/astro/blob/fd508a0fbb5148aafc180f1b14d3e47974777248/packages/db/src/core/load-file.ts#L55-L62


Instead of _needing_ to seed using a `seed.ts` file and passing that with `seedEntrypoint` maybe it could be possible to also use something like this to seed the db:
```ts
const Translations = defineTable({ foobar })
return {
  hooks: {
    "astro:db:ready": ({ db }) => {
      await db.insert(Translations).values({ footer })
    }
  }
}
```

**Comment by florian-lefebvre** [comment]
There's a similar issue with content collections because they're virtual imports so I'd love a solution to this issue in general

**Reply by Fryuni** [reply]
It is actually easier to do it for Astro DB than it is for Content Collections. It can be provided as a utility to use in nearly any hook.

**Comment by millette** [comment]
I'm not sure if this belongs here but I'd like to be able to draw db diagrams as hinted here:

https://github.com/drizzle-team/drizzle-orm/discussions/1480

**Comment by florian-lefebvre** [comment]
Astro DB has been deprecated
